src/ui/setup/dynamic/devtools.ts

```diff
@@ -106,39 +107,42 @@ export default async (store: Store) => {
     prefsService: getPrefsService(),
   };
 
-  extendStore(store, initialState, reducers, thunkArgs);
-
-  dbgClient.bootstrap(store);
-
-  // Initialize the socket so we can communicate with the server
-  initSocket(store, dispatchUrl);
-
-  addEventListener("Recording.uploadedData", (data: uploadedData) =>
-    store.dispatch(actions.onUploadedData(data))
-  );
-  addEventListener("Recording.awaitingSourcemaps", () =>
-    store.dispatch(actions.setAwaitingSourcemaps(true))
-  );
-  addEventListener("Recording.sessionError", (error: sessionError) =>
-    store.dispatch(
-      actions.setUnexpectedError({
-        ...error,
-        message: "Our apologies!",
-        content: "Something went wrong while replaying, we'll look into it as soon as possible.",
-        action: "refresh",
-      })
-    )
-  );
-
-  setupApp(store);
-  setupTimeline(store);
-  setupEventListeners(store);
-  setupGraphics(store);
-  initOutputSyntaxHighlighting();
-  setupMessages(store);
-  setupLogpoints(store);
-  setupExceptions(store);
-  setupReactDevTools(store);
+  if (!window.loaded) {
+    window.loaded = true;
```

Comment:
I think I'd do two things here:

1. extract to a function
2. more specific variable

```js
function bootstrapApp() {
  if (window.hasAlreadyBootstrapped) {
    return
  } 
    extendStore(store, initialState, reducers, thunkArgs);

    dbgClient.bootstrap(store);

    // Initialize the socket so we can communicate with the server
    initSocket(store, dispatchUrl);
}
```

Reply:
also we might need to include this as well

```js
  const settings = await getUserSettings();
  updateEnableRepaint(settings.enableRepaint);
  setupDemo();
```

Reply:
Yeah, actually, this whole function is supposed to be idempotent.

Reply:
I think